srcjs/init_shiny.js

```diff
@@ -332,13 +338,48 @@ function initShiny() {
   }
 
   $('.shiny-image-output, .shiny-plot-output, .shiny-report-theme').each(function() {
-    var id = getIdFromEl(this);
-    initialValues['.clientdata_output_' + id + '_bg'] = getComputedBgColor(this);
-    initialValues['.clientdata_output_' + id + '_fg'] = getStyle(this, "color");
-    initialValues['.clientdata_output_' + id + '_accent'] = getComputedLinkColor(this);
-    initialValues['.clientdata_output_' + id + '_font'] = getComputedFont(this);
+    var el = this, id = getIdFromEl(el);
+    initialValues['.clientdata_output_' + id + '_bg'] = getComputedBgColor(el);
+    initialValues['.clientdata_output_' + id + '_fg'] = getStyle(el, "color");
+    initialValues['.clientdata_output_' + id + '_accent'] = getComputedLinkColor(el);
+    initialValues['.clientdata_output_' + id + '_font'] = getComputedFont(el);
+    maybeAddThemeObserver(el);
   });
 
+
+  // Resend computed styles if this element's class or style changes which is
+  // especially important for invalidating getCurrentOutputInfo() when error
+  // messages modify the containers CSS in a non-welcome way.
+  // https://github.com/rstudio/shiny/issues/3196
+  //
+  // Unfortunately, won't solve getCurrentOutputInfo() not invalidating
+  // if this element's ancestors change in a meaningful way, but I don't
+  // see a reasonable way to solve that issue
+  // https://github.com/rstudio/shiny/issues/2998
+  function maybeAddThemeObserver(el) {
+    var cl = el.classList;
+    if (cl.contains("shiny-theme-observer")) return;
+
+    var reportTheme = cl.contains('shiny-image-output') || cl.contains('shiny-plot-output') || cl.contains('shiny-report-theme');
+    if (!reportTheme) return;
+
+    var observerCallback = new Debouncer(null, () => doSendTheme(el), 100);
+    var observer = new MutationObserver(() => observerCallback.normalCall());
+    var config = {attributes: true, attributeFilter: ['style', 'class']};
+    observer.observe(el, config);
+    cl.add("shiny-theme-observer"); // TODO: remove this class on unbind?
```

Comment:
If I'm understanding correctly, the `"shiny-theme-observer"` class exists only to indicate that there's a mutation observer. Instead of this class, I think it might make more sense to add a `.data()` value to the jQuery object.